plugins/auth-react/src/hooks/useCookieAuthRefresh/useCookieAuthRefresh.tsx

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+import { useEffect, useCallback, useMemo } from 'react';
+import {
+  discoveryApiRef,
+  fetchApiRef,
+  useApi,
+} from '@backstage/core-plugin-api';
+import { useAsync, useMountEffect } from '@react-hookz/web';
+import { ResponseError } from '@backstage/errors';
+
+/**
+ * @public
+ * A hook that will refresh the cookie when it is about to expire.
+ * @param options - Options for configuring the refresh cookie endpoint
+ */
+export function useCookieAuthRefresh(options: {
+  // The plugin id used for discovering the API origin
+  pluginId: string;
+  // The path used for calling the refresh cookie endpoint, default to '/cookie'
+  path?: string;
+}):
+  | { status: 'loading' }
+  | { status: 'error'; error: Error; retry: () => void }
+  | { status: 'success'; data: { expiresAt: string } } {
+  const { pluginId, path = '/cookie' } = options ?? {};
+  const fetchApi = useApi(fetchApiRef);
+  const discoveryApi = useApi(discoveryApiRef);
+
+  const channel = useMemo(() => {
+    return 'BroadcastChannel' in window
+      ? new BroadcastChannel(`${pluginId}-auth-cookie-expires-at`)
+      : null;
+  }, [pluginId]);
+
+  const [state, actions] = useAsync<{ expiresAt: string }>(async () => {
+    const apiOrigin = await discoveryApi.getBaseUrl(pluginId);
+    const requestUrl = `${apiOrigin}${path}`;
+    const response = await fetchApi.fetch(`${requestUrl}`, {
+      credentials: 'include',
+    });
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+    const data = await response.json();
+    if (!data.expiresAt) {
+      throw new Error('No expiration date found in response');
+    }
+    return data;
+  });
+
+  useMountEffect(actions.execute);
+
+  const retry = useCallback(() => {
+    actions.execute();
+  }, [actions]);
+
+  const refresh = useCallback(
+    (params: { expiresAt: string }) => {
+      // Randomize the refreshing margin with a margin of 1-4 minutes to avoid all tabs refreshing at the same time
+      // It cannot be less than 5 minutes otherwise the backend will return the same expiration date
+      const margin = (1 + 3 * Math.random()) * 60000;
+      const delay = Date.parse(params.expiresAt) - Date.now() - margin;
+      const timeout = setTimeout(retry, delay);
+      return () => clearTimeout(timeout);
+    },
+    [retry],
+  );
+
+  useEffect(() => {
+    // Only schedule a refresh if we have a successful response
+    if (state.status !== 'success' || !state.result) {
+      return () => {};
+    }
+    channel?.postMessage({
+      action: 'COOKIE_REFRESH_SUCCESS',
+      payload: state.result,
+    });
+    let cancel = refresh(state.result);
+    const listener = (
+      event: MessageEvent<{ action: string; payload: { expiresAt: string } }>,
+    ) => {
+      const { action, payload } = event.data;
+      if (action === 'COOKIE_REFRESH_SUCCESS') {
+        cancel();
+        cancel = refresh(payload);
```

Comment:
Should I remove the event listener here as well 🤔 ?

Reply:
No, just in the effect cleanup as you do atm afaik